pilot/pkg/serviceregistry/kube/controller/controller.go

```diff
@@ -376,6 +380,19 @@ func (c *Controller) onServiceEvent(pre, curr *v1.Service, event model.Event) er
 	// Create the standard (cluster.local) service.
 	svcConv := kube.ConvertService(*curr, c.opts.DomainSuffix, c.Cluster())
 
+	preNotVisible := isServiceNotVisible(pre)
+	currNotVisible := isServiceNotVisible(curr)
+	// If the service is consistently not visible, not process it.
+	// If the service changes from being visible to not visible, treat it as a delete event.
+	// If the service changes from being not visible to visible, treat it as an add event.
+	if preNotVisible && currNotVisible {
+		return nil
+	}
+	if currNotVisible {
+		c.deleteService(svcConv)
+		return nil
+	}
```

Comment:
we already have `serviceUpdateNeedsPush` function, can we put it there?

Reply:
The previous [commit](https://github.com/istio/istio/compare/8fc727f7436cfe124f41a1b1428f66fced493b7d..2c65b7ad90ee1337839b9826b3c811ef01bf560b) was like this. 

Made changes based on this [review](https://github.com/istio/istio/pull/47009#discussion_r1327015178). If don't want to cache in serviceMaps, maybe not enough to only modify serviceUpdateNeedsPush.

I'm not very sure which way is more suitable.

Reply:
I think @howardjohn means to protect with `ISTIO_ENABLE_OPTIMIZED_SERVICE_PUSH`.

however putting in `serviceUpdateNeedsPush` is not efficient

Reply:
I do want to cache it in serviceMaps, who knows what will break if we just decide to delete it when it really exists

Reply:
@hzxuzhonghu What do you think I roll back to the previous version?